modules/gapi/src/streaming/vpl/vpl_source_impl.cpp

```diff
@@ -0,0 +1,361 @@
+            }
+
+            // create session driving engine if required
+            if (!engine) {
+
+                std::unique_ptr<VPLAccelerationPolicy> acceleration = initializeHWAccel();
+                assert(mfx_impl_desription && "mfx_impl_desription must exist");
+
+                // TODO  Add factory stati method in VPLProcessingEngine
+                if (mfx_impl_desription->ApiVersion.Major >= VPL_NEW_API_MAJOR_VERSION)
+                {
+                    engine.reset(new VPLDecodeEngine(std::move(acceleration)));
+                }
+                else
+                {
+                    engine.reset(new VPLLegacyDecodeEngine(std::move(acceleration)));
+                }
+            }
+
+            //create decoder for session accoring to header recovered from source file
+            DecoderParams decoder_param = create_decoder_from_file(dec_it->second, source_handle.get());
+
+            // create engine session for processing mfx session pipeline
+            engine->initialize_session(mfx_session, std::move(decoder_param),
+                                                    std::move(source_handle));
+        } catch(const std::exception& ex) {
+            std::stringstream ss;
+            ss << ex.what() << ". Unload VPL session: " << mfx_session;
+            const std::string str = ss.str();
+            GAPI_LOG_WARNING(nullptr, str);
+            MFXClose(mfx_session);
+            util::throw_error(std::logic_error(str));
+        }
+    } catch (const std::exception& ex) {
+        std::stringstream ss;
+        ss << "Cannot create VPL source Impl, error: " << ex.what() << ". Unload MFX handle: " << mfx_handle;
+        const std::string str = ss.str();
+        GAPI_LOG_WARNING(nullptr, str);
+        MFXUnload(mfx_handle);
+        util::throw_error(std::logic_error(str));
+    }
+
+    //prepare
+    engine->process(mfx_session);
+}
+
+DecoderParams VPLSourceImpl::create_decoder_from_file(const CFGParamValue& decoder, FILE* source_ptr)
+{
+    if (!source_ptr) {
+        throw std::runtime_error("Cannot create decoder, source is nullptr");
+    }
+
+    mfxBitstream bitstream{};
+    const int BITSTREAM_BUFFER_SIZE = 2000000;
+    bitstream.MaxLength = BITSTREAM_BUFFER_SIZE;
+    bitstream.Data      = (mfxU8 *)calloc(bitstream.MaxLength, sizeof(mfxU8));
+    if(!bitstream.Data) {
+        throw std::runtime_error("Cannot allocate bitstream.Data bytes: " +
+                             std::to_string(bitstream.MaxLength * sizeof(mfxU8)));
+    }
+    bitstream.CodecId = decoder.Data.U32;
+
+    mfxStatus sts = ReadEncodedStream(bitstream, source_ptr);
+    if(MFX_ERR_NONE != sts) {
+        throw std::runtime_error("Error reading bitstream, error: " +
+                                 mfxstatus_to_string(sts));
+    }
+
+    // Retrieve the frame information from input stream
+    mfxVideoParam mfxDecParams {};
+    mfxDecParams.mfx.CodecId = decoder.Data.U32;
```

Comment:
What about other data types?

Reply:
`Data.U32` is some sort of `union` . Actual value type depends on parameter name
`Decoder...CodecId` according to VPL Documentation should be UINT32 here

Reply:
Please add a comment on that